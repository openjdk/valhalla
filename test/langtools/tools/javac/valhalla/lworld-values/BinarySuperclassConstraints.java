/*
 * @test /nodynamiccopyright/
 * @bug 8242900
 * @summary Verify various constraints for an inline class's BINARY super types.
 * @compile -XDrawDiagnostics -XDdev SuperclassCollections.java
 * @compile/fail/ref=BinarySuperclassConstraints.out -XDrawDiagnostics -XDdev BinarySuperclassConstraints.java
 */

public class BinarySuperclassConstraints {

    // -------------------------------------------------------------

    // Test that super class cannot be concrete, including express jlO
    inline class I0 extends SuperclassCollections.BadSuper {} // ERROR: concrete super class

    // Test that abstract class is allowed to be super including when extending jlO
    inline class I3 extends SuperclassCollections.GoodSuper implements SuperclassCollections.GoodSuperInterface {} // jlO can be indirect super class
    inline class I4 extends SuperclassCollections.Integer {}
    inline class I5 extends Number {
        public double doubleValue() { return 0; }
        public float floatValue() { return 0; }
        public long longValue() { return 0; }
        public int intValue() { return 0; }
    }

    // -------------------------------------------------------------

    // Test that super class cannot define instance fields.
    inline class I6 extends SuperclassCollections.SuperWithInstanceField_01 {} // ERROR:

    inline class I7 extends SuperclassCollections.SuperWithStaticField {} // OK.

    // -------------------------------------------------------------

    // Test that no-arg constructor must be empty
    inline class I8 extends SuperclassCollections.SuperWithEmptyNoArgCtor_02 {}

    inline class I9 extends SuperclassCollections.SuperWithNonEmptyNoArgCtor_01 {} // ERROR:

    inline class I10 extends SuperclassCollections.SuperWithArgedCtor_01 {} // ERROR:

    inline class I11 extends SuperclassCollections.SuperWithInstanceInit_01 {} // ERROR:

    inline class I12 extends SuperclassCollections.SuperWithSynchronizedMethod_1 {} // ERROR:

    inline class I13 extends SuperclassCollections.InnerSuper {}
}
