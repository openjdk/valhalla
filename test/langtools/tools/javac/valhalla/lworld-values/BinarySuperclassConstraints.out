BinarySuperclassConstraints.java:14:12: compiler.err.inline.type.must.not.implement.identity.object: BinarySuperclassConstraints.I0
BinarySuperclassConstraints.java:29:12: compiler.err.super.field.not.allowed: x, BinarySuperclassConstraints.I6, SuperclassCollections.SuperWithInstanceField
BinarySuperclassConstraints.java:38:12: compiler.err.super.no.arg.constructor.must.be.empty: SuperclassCollections.SuperWithNonEmptyNoArgCtor(), BinarySuperclassConstraints.I9, SuperclassCollections.SuperWithNonEmptyNoArgCtor
BinarySuperclassConstraints.java:40:12: compiler.err.super.constructor.cannot.take.arguments: SuperclassCollections.SuperWithArgedCtor(java.lang.String), BinarySuperclassConstraints.I10, SuperclassCollections.SuperWithArgedCtor
BinarySuperclassConstraints.java:42:12: compiler.err.super.no.arg.constructor.must.be.empty: SuperclassCollections.SuperWithInstanceInit(), BinarySuperclassConstraints.I11, SuperclassCollections.SuperWithInstanceInit
BinarySuperclassConstraints.java:44:12: compiler.err.super.method.cannot.be.synchronized: foo(), BinarySuperclassConstraints.I12, SuperclassCollections.SuperWithSynchronizedMethod
BinarySuperclassConstraints.java:46:12: compiler.err.encl.class.required: SuperclassCollections.InnerSuper
7 errors
